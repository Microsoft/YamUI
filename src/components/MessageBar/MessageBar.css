/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
@import 'src/css/variables/colors.css';

.y-message-bar {
  background-color: $interactiveElementBackround__focused;
  border: solid 0 $alertColor__info;
  border-radius: 1px;
  border-left-width: 4px;
  color: $textColor;
  padding: 7px 12px 8px;
}

.y-message-bar--actions {
  white-space: nowrap;
}

.y-message-bar__type-warning {
  border-color: $alertColor__warning;
}

.y-message-bar__type-error {
  border-color: $alertColor__error;
}
