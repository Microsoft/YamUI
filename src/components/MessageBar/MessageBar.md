### Notes for use

Text that is too long for the context will wrap to two or more lines, but you should aim for short strings. Actions will grow from the right but should also take up as little horizontal space as possible.

### Examples

Informational:

```js { "props": { "data-description": "basic" } }
<MessageBar>
  Something you should know about.
</MessageBar>
```

Warning:

```js { "props": { "data-description": "with type" } }
const { MessageBarType } = require('.');

<MessageBar type={MessageBarType.WARNING}>
  Watch out, something unexpected might happen.
</MessageBar>
```

Warning with an icon action:

```js { "props": { "data-description": "with icon action" } }
const { IconSize } = require('../Icon');
const Lock = require('../Icon/icons/Lock').default;
const { TextSize } = require('../Text');
const { MessageBarType } = require('.');

const action = (
  <span>
    <Lock size={IconSize.XSMALL} />
    <Text size={TextSize.SMALL}>+1</Text>
  </span>
);

<MessageBar actions={action} type={MessageBarType.WARNING}>
  Watch out, something unexpected might happen.
</MessageBar>
```

Warning with an icon action and multiple lines:

```js { "props": { "data-description": "with icon action" } }
const { IconSize } = require('../Icon');
const Lock = require('../Icon/icons/Lock').default;
const { TextSize } = require('../Text');
const { MessageBarType } = require('.');

const action = (
  <span>
    <Lock size={IconSize.XSMALL} />
    <Text size={TextSize.SMALL}>+1</Text>
  </span>
);

<MessageBar actions={action} type={MessageBarType.WARNING}>
  Watch out, something unexpected might happen. This is an example on how a pretty long copy is going to scale. As you can see it keeps the proper margin between the main copy text and the icon on the right.
</MessageBar>
```

Error with a link action:

```js { "props": { "data-description": "with link action" } }
const { MessageBarType } = require('.');

const link = <Clickable>Retry</Clickable>;

<MessageBar actions={link} type={MessageBarType.ERROR}>
  Oh no, something bad happened and we couldn't complete your action!
</MessageBar>
```
