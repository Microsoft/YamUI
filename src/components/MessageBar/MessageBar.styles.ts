/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import { palette, semanticColors } from '../../util/colors';
import { mergeStyleSets } from '@uifabric/styling';
import { memoizeFunction } from '@uifabric/utilities';
import { MessageBarType } from './types';

export interface MessageBarStyleProps {
  type: MessageBarType;
}

const borderColors = {
  [MessageBarType.WARNING]: semanticColors.warningBackground,
  [MessageBarType.ERROR]: semanticColors.errorBackground,
  [MessageBarType.INFO]: palette.neutralTertiaryAlt,
};

export const getClassNames = memoizeFunction((styleProps: MessageBarStyleProps) => {
  const { type } = styleProps;

  return mergeStyleSets({
    root: {
      backgroundColor: palette.neutralLighter,
      borderColor: borderColors[type],
      borderRadius: '1px',
      borderLeftStyle: 'solid',
      borderLeftWidth: '4px',
      padding: '7px 12px 8px',
    },
    actions: {
      whiteSpace: 'nowrap',
    },
  });
});
