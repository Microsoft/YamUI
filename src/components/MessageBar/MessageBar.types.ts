/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import { NestableBaseComponentProps } from '../../util/BaseComponent/props';

export enum MessageBarType {
  INFO = 'info',
  WARNING = 'warning',
  ERROR = 'error',
}

export interface MessageBarProps extends NestableBaseComponentProps {
  /**
   * Type of message being displayed.
   * @default MessageBarType.INFO
   */
  type?: MessageBarType;

  /**
   * Elements to be displayed as actions for this message.
   */
  actions?: React.ReactNode;
}
