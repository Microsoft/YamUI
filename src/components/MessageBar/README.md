# MessageBar

Renders a component that displays relevant status information. You can use a MessageBar to tell the user about a situation and optionally provide actions for them to take.


## Properties

| Prop | Type | Description |
| ---- | ---- | ----------- |
| ariaLabel * | string | A description of the message bar for the benefit of screen readers. |
| type | MessageBarType | Dictates the color scheme of the component. Defaults to `MessageBarType.INFO` |
| actions | ReactNode | A component to render in the actions gutter. Defaults to `null` |

\* indicates property is required

## Notes for use

Text that is too long for the context will wrap to two or more lines but you should aim for short strings. Actions will grow from the right but should also take up as little horizontal space as possible. 

## Examples

### Basic usage

```js
const myIconComponent = <Icon icon="lock" />
<MessageBar
  ariaLabel="Warning that your post will be visible to some outside your network"
  type={MessageBarType.WARNING}
  actions={myIconComponent}
>
  <Text>Your post will be visible to some users outside your network.</Text>
</MessageBar>
```
