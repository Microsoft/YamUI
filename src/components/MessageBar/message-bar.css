/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
@import "src/css/variables/colors.css";
@import "src/css/variables/fonts.css";

.y-message-bar {
  display: flex;
  border-radius: 1px;
  border-left-width: 4px;
  border-left-style: solid;
  background-color: $color__popRock;
  padding: 7px 12px 8px 12px;
  font-size: $fontSize__small;
  line-height: $lineHeight__small;
}

.y-message-bar--message {
  flex-grow: 1;
  color: $color__blackMetal;
}

.y-message-bar--actions {
  padding-left: 32px;
  flex-grow: 0;
}

.y-message-bar__type-info {
  border-left-color: $messagebarBarColor__info;
}

.y-message-bar__type-warning {
  border-left-color: $messagebarBarColor__warning;
}

.y-message-bar__type-error {
  border-left-color: $messagebarBarColor__error;
}
