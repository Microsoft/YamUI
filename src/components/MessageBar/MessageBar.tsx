/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import '../../yamui';
import * as React from 'react';
import { join } from '../../util/classNames';
import { NestableBaseComponentProps } from '../../util/BaseComponent/props';
import { MessageBarType } from './types';
import Block, { TextSize } from '../Block';
import { FixedGridRow, FixedGridColumn, GutterSize } from '../FixedGrid';
import { getClassNames } from './MessageBar.styles';

export { MessageBarType };

export interface MessageBarProps extends NestableBaseComponentProps {
  /**
   * Type of message being displayed.
   * @default MessageBarType.INFO
   */
  type?: MessageBarType;

  /**
   * Elements to be displayed as actions for this message.
   */
  actions?: React.ReactNode;
}

/**
 * A `MessageBar` displays relevant status information. You can use a `MessageBar` to tell the user
 * about a situation, and optionally provide actions for them to take.
 */
export default class MessageBar extends React.Component<MessageBarProps> {
  public render() {
    const { actions, children, className, type = MessageBarType.INFO } = this.props;
    const classNames = getClassNames({ type });

    return (
      <Block textSize={TextSize.SMALL} className={join(['y-message-bar', className, classNames.root])}>
        <FixedGridRow gutterSize={GutterSize.XXLARGE}>
          <FixedGridColumn>{children}</FixedGridColumn>
          {actions && (
            <FixedGridColumn fixed={true} className={classNames.actions}>
              {actions}
            </FixedGridColumn>
          )}
        </FixedGridRow>
      </Block>
    );
  }
}
