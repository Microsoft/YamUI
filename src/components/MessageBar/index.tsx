/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import '../../yamui';
import * as React from 'react';
import classNames = require('classnames');
import { NestableBaseComponentProps } from '../../util/BaseComponent/props';
import { MessageBarType } from './enums';
import './message-bar.css';

export { MessageBarType };

export interface MessageBarProps extends NestableBaseComponentProps {
  ariaLabel: string;
  type?: MessageBarType;
  actions?: React.ReactNode;
  children: React.ReactNode;
}

const getActions = (props: MessageBarProps) => {
  if (!props.actions) {
    return null;
  }
  return <div className="y-message-bar--actions">{props.actions}</div>;
};

const MessageBar: React.StatelessComponent<MessageBarProps> = (props) => {
  const actions = getActions(props);
  const classes = classNames(`y-message-bar y-message-bar__type-${props.type}`, props.className);
  return (
    <div className={classes} aria-label={props.ariaLabel}>
      <div className="y-message-bar--message">{props.children}</div>
      {actions}
    </div>
  );
};

export default MessageBar;

MessageBar.defaultProps = {
  type: MessageBarType.INFO,
};
