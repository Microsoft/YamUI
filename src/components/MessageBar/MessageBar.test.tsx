/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import MessageBar, { MessageBarType, MessageBarProps } from './index';
import FakeLink from '../../components/FakeLink';

describe('<MessageBar />', () => {
  let component: ShallowWrapper<MessageBarProps, {}>;

  describe('with default options', () => {
    beforeEach(() => {
      component = shallow(<MessageBar>content</MessageBar>);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with additional className', () => {
    beforeEach(() => {
      component = shallow(<MessageBar className="TEST_CLASSNAME">content</MessageBar>);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with a type', () => {
    beforeEach(() => {
      component = shallow(<MessageBar type={MessageBarType.WARNING}>content</MessageBar>);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with actions', () => {
    beforeEach(() => {
      const link = <FakeLink>Link</FakeLink>;
      component = shallow(<MessageBar actions={link}>content</MessageBar>);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });
});
