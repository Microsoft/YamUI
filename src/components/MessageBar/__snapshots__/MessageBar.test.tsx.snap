// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MessageBar /> with a type matches its snapshot 1`] = `
<Block
  className=
      y-message-bar
      {
        background-color: rgb(243,245,248);
        border-color: rgb(255,197,108);
        border-left-width: 4px;
        border-radius: 1px;
        border: solid 0 rgb(168,176,189);
        color: rgb(52,58,65);
        padding-bottom: 8px;
        padding-left: 12px;
        padding-right: 12px;
        padding-top: 7px;
      }
  textSize="small"
>
  <FixedGridRow
    gutterSize="xxLarge"
  >
    <FixedGridColumn>
      content
    </FixedGridColumn>
  </FixedGridRow>
</Block>
`;

exports[`<MessageBar /> with actions matches its snapshot 1`] = `
<Block
  className=
      y-message-bar
      {
        background-color: rgb(243,245,248);
        border-left-width: 4px;
        border-radius: 1px;
        border: solid 0 rgb(168,176,189);
        color: rgb(52,58,65);
        padding-bottom: 8px;
        padding-left: 12px;
        padding-right: 12px;
        padding-top: 7px;
      }
  textSize="small"
>
  <FixedGridRow
    gutterSize="xxLarge"
  >
    <FixedGridColumn>
      content
    </FixedGridColumn>
    <FixedGridColumn
      className=

          {
            white-space: nowrap;
          }
      fixed={true}
    >
      <FakeLink>
        Link
      </FakeLink>
    </FixedGridColumn>
  </FixedGridRow>
</Block>
`;

exports[`<MessageBar /> with additional className matches its snapshot 1`] = `
<Block
  className=
      y-message-bar
      TEST_CLASSNAME
      {
        background-color: rgb(243,245,248);
        border-left-width: 4px;
        border-radius: 1px;
        border: solid 0 rgb(168,176,189);
        color: rgb(52,58,65);
        padding-bottom: 8px;
        padding-left: 12px;
        padding-right: 12px;
        padding-top: 7px;
      }
  textSize="small"
>
  <FixedGridRow
    gutterSize="xxLarge"
  >
    <FixedGridColumn>
      content
    </FixedGridColumn>
  </FixedGridRow>
</Block>
`;

exports[`<MessageBar /> with default options matches its snapshot 1`] = `
<Block
  className=
      y-message-bar
      {
        background-color: rgb(243,245,248);
        border-left-width: 4px;
        border-radius: 1px;
        border: solid 0 rgb(168,176,189);
        color: rgb(52,58,65);
        padding-bottom: 8px;
        padding-left: 12px;
        padding-right: 12px;
        padding-top: 7px;
      }
  textSize="small"
>
  <FixedGridRow
    gutterSize="xxLarge"
  >
    <FixedGridColumn>
      content
    </FixedGridColumn>
  </FixedGridRow>
</Block>
`;
